hazelcast/src/test/java/com/hazelcast/spi/hashslot/HashSlotArrayImplTest.java

```diff
@@ -0,0 +1,286 @@
+package com.hazelcast.spi.hashslot;
+
+import com.hazelcast.internal.memory.MemoryAccessor;
+import com.hazelcast.memory.HeapMemoryManager;
+import com.hazelcast.memory.MemoryManager;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.util.Arrays;
+import java.util.Random;
+
+import static com.hazelcast.memory.MemoryAllocator.NULL_ADDRESS;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(QuickTest.class)
+public class HashSlotArrayImplTest {
+
+    private static final int VALUE_LENGTH = 32;
+
+    private final Random random = new Random();
+    private MemoryManager memMgr;
+    private MemoryAccessor mem;
+    private HashSlotArray hsa;
+
+    @Before
+    public void setUp() throws Exception {
+        memMgr = new HeapMemoryManager(32 << 20);
```

Comment:
Memory leaks and invalid `free()` calls are my worry, too; however if we test for them on the EE side, then a broken OSS PR will be merged as green, followed by failing EE PRs. We need to test the code at its origin. Therefore my proposal is to improve the test allocator to validate memory management. `free()` calls can be managed as you showed and memory leak can be tested for by introducing a simple allocated memory counter. In the future we'll hopefully introduce `MemoryStatistics` to the general `MemoryManager`.